src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -252,12 +262,26 @@ class CorrectUmis
 
   /** Given a UMI sequence and a set of fixed UMIs, report the best match. */
   private[umi] def findBestMatch(bases: String, umis: Array[String]): UmiMatch = {
-    val mismatchLimit = this.maxMismatches + minDistance
-    val mismatches    = umis.map(umi => Sequences.countMismatches(bases, umi))
-    val min           = mismatches.min
-    val matched       = (min <= maxMismatches) && (mismatches.count(m => m < min + this.minDistance) == 1)
-
-    UmiMatch(matched, umis(mismatches.indexOf(min)), min)
+    if (cacheSize > 0 && cache.containsKey(bases)) cache.get(bases)
+    else {
+      val mismatches    = umis.map(umi => Sequences.countMismatches(bases, umi))
+      val min           = mismatches.min
+      val matched       = (min <= maxMismatches) && (mismatches.count(m => m < min + this.minDistance) == 1)
+      val umiMatch      = UmiMatch(matched, umis(mismatches.indexOf(min)), min)
+      if (cacheSize > 0) cache.put(bases, umiMatch)
+      umiMatch
+    }
   }
 }
 
+/** Least recently used cache */
+private class LRUCache[K, V](maxEntries: Int)
+  extends java.util.LinkedHashMap[K, V](100, .75f, true) {
+  override def removeEldestEntry(eldest: java.util.Map.Entry[K, V]): Boolean = size > maxEntries
+}
+
+private object LRUCache {
+  def apply[K, V](maxEntries: Int): mutable.Map[K, V] = {
+    synchronizedMap(new LRUCache[K, V](maxEntries)).asScala
+  }
+}
```

Comment:
I think:

1. This should use scala.collection.mutable.LinkedHashMap
2. It should not extend, but have the `map` as a `private` val
3. Only public methods should be `get()` and `put()`, both synchronized, with `put()` doing the eviction as needed.

Reply:
You could also add a `getOrElseUpdate(key: Key, value: => Value)`